File: 494-TargetSum/494-TargetSum.cpp
// Last updated: 12/8/2025, 11:28:38 PM
class Solution {
public:
    int total = 0;
    int findTargetSumWays(vector<int>& nums, int target) {
        total = accumulate(nums.begin(), nums.end(), 0);
        int n = nums.size();
        vector<vector<int>> memo(n, vector<int>(2 * total + 1, numeric_limits<int>::min()));
        return calculate(nums, 0, 0, target, memo);
    }

    int calculate(const vector<int>& nums, int i, int cur, int target, vector<vector<int>>& memo) {
        if (i == nums.size()) {
            if (cur == target) {
                return 1;
            } else {
                return 0;
            }
        }
        if (memo[i][cur + total] != numeric_limits<int>::min()) {
            return memo[i][cur + total];
        }

        int add = calculate(nums, i + 1, cur + nums[i], target, memo);
        int sub = calculate(nums, i + 1, cur - nums[i], target, memo);
        memo[i][cur + total] = add + sub;
        return add + sub;
    }
};

File: 494-TargetSum/494-TargetSum.py
# Last updated: 12/8/2025, 11:28:35 PM
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        total = sum(nums)
        n = len(nums)
        dp = [[0] * (2 * total + 1) for _ in range(n)]
        dp[0][total + nums[0]] = 1
        dp[0][total - nums[0]] += 1  # could be same, so +=
        for i in range(1, n):
            for j in range(-total, total + 1):
                if dp[i - 1][j + total] > 0:
                    dp[i][j + nums[i] + total] += dp[i - 1][j + total]
                    dp[i][j - nums[i] + total] += dp[i - 1][j + total]
        return 0 if abs(target) > total else dp[n - 1][target + total]
